### PR TITLE
DCOS-21981 [data-service package] Graphql Library

### PR DESCRIPTION
**Pull request by Fabs**
This is the Graphql library implementation needed for the future data-service package. 

TO TEST
- Read test cases and implementation.
- If you feel something is not covered for queries, you can play with the testcases. (have in mind that this is not a full implementation)

TODO
- [x] Implement a minimal Graphql library
- [x] Add package structure

### CODE REVIEW

File: packages/data-service/src/graphqlObservable.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import { Observable } from "rxjs/Observable";
+
+// TODO: refactor this once it has a more compreehensive implementation
```
Comment (Fabs) [inline packages/data-service/src/graphqlObservable.js]: https://jira.mesosphere.com/browse/DCOS-21853

File: packages/data-service/package.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "data-service",
+  "module": "./src/index.js",
+  "version": "0.0.1",
+  "license": "Apache-2.0",
+  "dependencies": {
+    "graphql": "0.13.2",
```
Comment (Fabs) [inline packages/data-service/package.json]: @nLight after moving to the structure you asked, I ran into a problem I don't know how to solve. Those dependencies do not get installed when I do npm install on the original repository. I tried peerDependencies briefly was not able to solve it.

If you could please take a look that would be awesome.

File: packages/data-service/src/__tests__/graphqlObservable-test.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        query roscosmos {
+          launched {
+            name
+          }
+        }
+      `;
+
+      const expectedData = [{ name: "discovery", firstFlight: 1984 }];
+      const dataSource = Observable.of(expectedData);
+      const expected = m.cold("(a|)", {
```
Comment (DanielMSchmidt) [inline packages/data-service/src/__tests__/graphqlObservable-test.js]: Would it be hard to add a test for changing data and not "only" one-time data being exposed?

File: packages/data-service/src/graphqlObservable.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+// of the graphql api
+
+// WARNING: This is NOT a spec complete graphql implementation
+// https://facebook.github.io/graphql/October2016/
+
+// eslint-disable-next-line import/prefer-default-export
+export const graphqlObservable = (doc, schema, context) => {
```
Comment (DanielMSchmidt) [inline packages/data-service/src/graphqlObservable.js]: would you mind making this a named function? It is quite big and I would love to see it's name on a stack trace.

File: packages/data-service/src/graphqlObservable.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  const translateOperation = {
+    query: "Query"
+  };
+
+  const resolveStep = (typeMap, definition, context, parent) => {
```
Comment (DanielMSchmidt) [inline packages/data-service/src/graphqlObservable.js]: I think this should also be a named function, it seems pretty important

File: packages/data-service/src/graphqlObservable.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    if (definition.kind === "OperationDefinition") {
+      const nextTypeMap = typeMap[
```
Comment (DanielMSchmidt) [inline packages/data-service/src/graphqlObservable.js]: This seems brittle, there are 4 null checks missing, do we refactor this later? TS could help here (at least me to not do dump comments like this ;) )

File: packages/data-service/src/graphqlObservable.js (hunk 4 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: totally. I think I was caught up in the style but it should have a name definitely.

File: packages/data-service/src/graphqlObservable.js (hunk 5 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: indeed.

File: packages/data-service/src/__tests__/graphqlObservable-test.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+    itMarbles("resolve multiple queries with alias", function(m) {
+      const query = gql`
+        query nasa {
+          launched {
+            name
+          }
+        }
+
+        query roscosmos {
```
Comment (nLight) [inline packages/data-service/src/__tests__/graphqlObservable-test.js]: ❤️

File: packages/data-service/src/graphqlObservable.js (hunk 5 above)
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: any particular this couldn't be a top level function?

File: packages/data-service/src/graphqlObservable.js (hunk 6 above)
Comment (nLight) [reply packages/data-service/src/graphqlObservable.js]: `graphql-tools` ✅ 
`graphql-tag` :trollface:

File: packages/data-service/src/graphqlObservable.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+// https://facebook.github.io/graphql/October2016/
+
+// eslint-disable-next-line import/prefer-default-export
+export const graphqlObservable = (doc, schema, context) => {
+  const translateOperation = {
+    query: "Query"
+  };
+
+  const resolveStep = (typeMap, definition, context, parent) => {
+    if (definition.kind === "OperationDefinition") {
```
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: As discussed we could do a `pipe` here and test all the steps independently

File: packages/data-service/src/graphqlObservable.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+        return acc.combineLatest(resolvedObservable, merger);
+      }, Observable.of({}));
+    }
+
+    // Node Field
+    if (definition.kind === "Field" && definition.selectionSet !== undefined) {
+      const args = definition.arguments
+        .map(arg => {
+          if (arg.value.kind === "Variable") {
```
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: I'd maybe even prefer a ternary here:
```js
definition.arguments.map(
  arg =>
    arg.value.kind === "Variable"
      ? { [arg.name.value]: context[arg.value.name.value] }
      : { [arg.name.value]: arg.value.value }
);
```

File: packages/data-service/src/graphqlObservable.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  if (doc.definitions.length === 1) {
+    return resolveStep(schema._typeMap, doc.definitions[0], context, null);
+  }
+
+  return doc.definitions.reduce((acc, definition) => {
+    const resolvedObservable = resolveStep(
+      schema._typeMap,
+      definition,
+      context,
+      null
```
Comment (nLight) [inline packages/data-service/src/graphqlObservable.js]: Sometimes we omit the argument, sometimes we pass `null` explicitly, I struggle to see a reason.

File: packages/data-service/src/graphqlObservable.js (hunk 1 above)
Comment (nLight) [reply packages/data-service/src/graphqlObservable.js]: 🔤 compreehensive -> comprehensive

File: packages/data-service/package.json [hunk 11]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "data-service",
+  "module": "./src/index.js",
+  "version": "0.0.1",
```
Comment (nLight) [inline packages/data-service/package.json]: usually semver doesn't start with a bugfix but it's ok :)

Comment (Fabs) [reply packages/data-service/package.json]: Ubuntu first bug:
https://bugs.launchpad.net/ubuntu/+bug/1

File: packages/data-service/src/graphqlObservable.js (hunk 1 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: I have a spell checker on intelij, need to love him more :D

File: packages/data-service/src/graphqlObservable.js (hunk 6 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: Could we move forward by saying that we will address that as we go:
https://jira.mesosphere.com/browse/DCOS-22006

To the point of trusting our implementation, I would also like to bring the tests form the graphql library into our implementation.

File: packages/data-service/src/graphqlObservable.js (hunk 8 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: I love the idea, but can we postpone that one when we do this?
https://jira.mesosphere.com/browse/DCOS-22007

File: packages/data-service/src/graphqlObservable.js (hunk 9 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: ✅

File: packages/data-service/src/graphqlObservable.js (hunk 10 above)
Comment (Fabs) [reply packages/data-service/src/graphqlObservable.js]: I think in this case it acts a little as a reminder that this exists. I don't know the answer to that one as well, especially being the last argument.

File: packages/data-service/package.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "data-service",
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/data-service/package.json]: Thank you for explaining this 👍

File: packages/data-service/src/__tests__/graphqlObservable-test.js (hunk 3 above)
Comment (DanielMSchmidt) [reply packages/data-service/src/__tests__/graphqlObservable-test.js]: Could you do it maybe in a separate PR? Would love to see it 👍

Comment (Fabs) [reply packages/data-service/src/__tests__/graphqlObservable-test.js]: Yes. I will create a ticket to do that. I think it is very valuable. There is an approach to copy the  tests from `graphql-js` so we can be sure we are as good as they are.

**Pull request merged**